### PR TITLE
Start and stop folders (fixes #215)

### LINKED ISSUE #215: Per folder "pause sync" toggle
When sync is paused, no data should be transmitted.

There should also be a Rest API for this (pause and resume).
#192

Edit: I think there should also a `-pause` command line flag to launch in paused mode (and/or remember the state on shutdown).

### PR DESCRIPTION
**Pull request by calmh**
Here's one to pull an old issue out of it's grave and kill it again... This adds the capability of starting and stopping folders (i.e. pausing them). Any hidden dangers here apart from the usual? Starting/resuming a folder causes a new initial scan, so any changes done in the meantime should cause a conflict when necessary and so on.

This is also the base for just restarting a folder on any config changes to it.

### CODE REVIEW

File: cmd/syncthing/gui.go [hunk 1]
```diff
@@ -169,6 +169,8 @@ func (s *apiSvc) Serve() {
 	postRestMux.HandleFunc("/rest/system/restart", s.postSystemRestart)        // -
 	postRestMux.HandleFunc("/rest/system/shutdown", s.postSystemShutdown)      // -
 	postRestMux.HandleFunc("/rest/system/upgrade", s.postSystemUpgrade)        // -
+	postRestMux.HandleFunc("/rest/system/stop", s.postSystemStopFolder)        // folder
+	postRestMux.HandleFunc("/rest/system/start", s.postSystemStartFolder)      // folder
```
Comment (calmh) [inline cmd/syncthing/gui.go]: Debatable if this goes under /rest/system or /rest/db. We have /rest/db/scan, and we're not stopping the entire *system*, but it's not strictly speaking a db operation...

**Comment by calmh** [comment]
@st-jenkins test this please

**Comment by AudriusButkevicius** [comment]
I'd expect not to incurre network traffic for a stopped folder.

**Comment by calmh** [comment]
How would we implement that? Drop all connection and reconnect, making sure it's not present in the clusterconfig message?

**Comment by uok** [comment]
Thanks for working on this, this will make many people happy :smile: 
My suggestion: to avoid confusion name it "pause" and "start" because if I remember correctly there is also a stopped folder state if something is wrong with the folder

**Comment by calmh** [comment]
I did that to start with, but technically the state is exactly the same - "not running". We could call it paused in the error case as well possibly. One can now, as a side effect, force start an errored folder as well. (It'll immediately stop again if the error remains, though.)

**Comment by uok** [comment]
yeah, something to differentiate "this folder is not working because something is wrong" vs. "you paused it earlier" would be helpful

**Comment by calmh** [comment]
There is a differentiator in the GUI though, the presence or absence of an error message. We could use that to set the color appropriately too - grey stopped vs red stopped?

**Comment by uok** [comment]
As long as it is clear that one is an error and the other is because of user I'm happy. Although I prefer different wordings "paused" and "stopped" or "stopped" and "error" to avoid playing roulette with colorblind people

**Comment by AudriusButkevicius** [comment]
I mean stopping responding to Request, not sure how you'd do that.

**Comment by calmh** [comment]
@AudriusButkevicius Ah, that we should certainly do. In this case, just checking in Request() that there is a runner would be sufficient.

@uok 

screen shot 2015-08-23 at 20 10 44

vs

screen shot 2015-08-23 at 20 10 50

?

**Comment by uok** [comment]
Thanks for the mockup, I think just "error" in red is enough. Then the user expands the folder and sees "folder marker missing" and knows what to do

**Comment by calmh** [comment]
Not a mockup. :) I went with Paused as opposed to Stopped in the error case, it's cleaner.

**Comment by AudriusButkevicius** [comment]
The problem with responding to Request with DoestNotExist, is that you could potentially get all requests just because you have 0 blocks in flight, potentially consuming a fair amount of bandwidth.

**Comment by calmh** [comment]
We *could* send an empty Index, thus effectively saying we have no files, leave us alone. We then need to resend the Index when we start the folder, which in itself can be quite expensive, but hey, we're supposed to be in a state where we can consume bandwidth then right?

**Comment by calmh** [comment]
(Those event test timeouts are annoying... I'm pretty sure it's just the build server being slow when it's loaded with other builds, but still...)

**Comment by AudriusButkevicius** [comment]
Yeah but this breaks with persistent indexes.

**Comment by calmh** [comment]
Not really... It's ugly, as we clear out the index on the other side unnecessarily and then need to resend it. Perhaps there actually needs to be a "pause folder" message that blocks other nodes from sending traffic to us for a given folder...

**Comment by calmh** [comment]
Or, as you were thinking about earlier, permitting new ClusterConfigs on the fly. Then we could just remove the folder from the announced list... But this requires more plumbing for sure.

**Comment by AudriusButkevicius** [comment]
Anyway, I think this would be an expectation with this feature (not consuming bandwidth while its stopped), hence we should address it now.

**Comment by calmh** [comment]
Yeah, you're right.

**Comment by AudriusButkevicius** [comment]
Whichever way we decide to do that though, I don't really care. We disconnect reconnect as part of adding folders, so empty index is not a big deal.

**Comment by Getron** [comment]
Great, thank you devs for working on this, this will make me, and many other people out there very happy to have this on Syncthing.

Just a proposal: red color for "Stopped", is just little scary, it seems like it is an error, or somthing wrong, but, in fact, it is just stopped folder, or device.

Should maybe "Stopped" have other color, like, maybe, blue, or orange, or maybe, light gray. What you mean?

**Comment by uok** [comment]
@calmh : updated screenshot looks perfect :+1: 
@Getron : red is the correct color - the folder is stopped because of an error (e.g. user moved shared folder and Syncthing cannot find it) as opposed to paused which was done by the user

**Comment by calmh** [comment]
Yeah #2188 is more evolved, this will get the tweaked UI once I figure out how to handle the concerns above. :)

**Comment by calmh** [comment]
@st-jenkins test this please

**Pull request closed without merging**